src/zkp/whitelist.rs

```diff
@@ -0,0 +1,470 @@
+    }
+
+    /// Obtains a raw mutable pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::WhitelistSignature {
+        &mut self.0
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::LowerHex for WhitelistSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in self.serialize().iter() {
+            write!(f, "{:02x}", ch)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::Display for WhitelistSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::Debug for WhitelistSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+#[cfg(feature = "std")]
+impl str::FromStr for WhitelistSignature {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<WhitelistSignature, Error> {
+        let mut buf = vec![0; s.len() / 2];
+        from_hex(s, &mut buf).map_err(|_| Error::InvalidWhitelistSignature)?;
+        WhitelistSignature::from_slice(&buf)
+    }
+}
+
+#[cfg(all(feature = "serde", feature = "std"))]
+impl ::serde::Serialize for WhitelistSignature {
+    fn serialize<S: ::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        if s.is_human_readable() {
+            s.collect_str(self)
+        } else {
+            s.serialize_bytes(&self.serialize())
+        }
+    }
+}
+
+#[cfg(all(feature = "serde", feature = "std"))]
+impl<'de> ::serde::Deserialize<'de> for WhitelistSignature {
+    fn deserialize<D: ::serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        use serde_util;
+
+        if d.is_human_readable() {
+            d.deserialize_str(serde_util::FromStrVisitor::new("an ASCII hex string"))
+        } else {
+            d.deserialize_bytes(serde_util::BytesVisitor::new(
+                "a bytestring",
+                WhitelistSignature::from_slice,
+            ))
+        }
+    }
+}
+
+impl CPtr for WhitelistSignature {
+    type Target = ffi::WhitelistSignature;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+#[cfg(all(test, feature = "global-context"))]
+mod tests {
+    use super::*;
+    use rand::thread_rng;
+    use SECP256K1;
+
+    fn test_whitelist_proof_roundtrip(n_keys: usize) {
+        let mut rng = thread_rng();
+        let keys_online = (0..n_keys)
+            .map(|_| SECP256K1.generate_keypair(&mut rng))
+            .collect::<Vec<_>>();
+        let keys_offline = (0..n_keys)
+            .map(|_| SECP256K1.generate_keypair(&mut rng))
+            .collect::<Vec<_>>();
+
+        // extract the pubkeys to serve as PAK list
+        let pak_online = keys_online.iter().map(|p| p.1).collect::<Vec<_>>();
+        let pak_offline = keys_offline.iter().map(|p| p.1).collect::<Vec<_>>();
```

Comment:
[`unzip`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.unzip) might be useful here.

Reply:
Fancy, updated.